File: backend/src/services/recurrence_service.py
from typing import Optional
import uuid
from datetime import datetime
from sqlmodel import Session
from src.models.recurrence_pattern import RecurrencePattern, RecurrencePatternCreate, RecurrencePatternUpdate
from src.repositories.recurrence_repository import RecurrenceRepository


class RecurrenceService:
    def __init__(self):
        self.repository = RecurrenceRepository()

    def create_recurrence_pattern(self, session: Session, recurrence_create: RecurrencePatternCreate) -> RecurrencePattern:
        """
        Create a new recurrence pattern
        """
        return self.repository.create_recurrence_pattern(session, recurrence_create)

    def get_recurrence_pattern_by_id(self, session: Session, recurrence_id: uuid.UUID) -> Optional[RecurrencePattern]:
        """
        Get a recurrence pattern by ID
        """
        return self.repository.get_recurrence_pattern_by_id(session, recurrence_id)

    def update_recurrence_pattern(
        self, 
        session: Session, 
        recurrence_id: uuid.UUID, 
        recurrence_update: RecurrencePatternUpdate
    ) -> Optional[RecurrencePattern]:
        """
        Update a recurrence pattern
        """
        return self.repository.update_recurrence_pattern(session, recurrence_id, recurrence_update)

    def delete_recurrence_pattern(self, session: Session, recurrence_id: uuid.UUID) -> bool:
        """
        Delete a recurrence pattern
        """
        return self.repository.delete_recurrence_pattern(session, recurrence_id)

    def validate_recurrence_pattern(self, recurrence_data: RecurrencePatternCreate) -> bool:
        """
        Validate recurrence pattern data
        """
        # Validate frequency
        if recurrence_data.frequency not in ["daily", "weekly", "monthly", "yearly"]:
            return False
            
        # Validate interval
        if recurrence_data.interval < 1:
            return False
            
        # Validate days of week if frequency is weekly
        if recurrence_data.frequency == "weekly" and recurrence_data.days_of_week:
            valid_days = {"mon", "tue", "wed", "thu", "fri", "sat", "sun"}
            if not all(day in valid_days for day in recurrence_data.days_of_week):
                return False
                
        # Validate day of month if frequency is monthly
        if recurrence_data.frequency == "monthly" and recurrence_data.day_of_month:
            if not (1 <= recurrence_data.day_of_month <= 31):
                return False
                
        # Validate end condition
        if recurrence_data.end_condition not in ["never", "after", "on_date"]:
            return False
            
        # Validate end after occurrences if end condition is "after"
        if recurrence_data.end_condition == "after" and recurrence_data.end_after_occurrences:
            if recurrence_data.end_after_occurrences < 1:
                return False
                
        # Validate end date if end condition is "on_date"
        if recurrence_data.end_condition == "on_date" and recurrence_data.end_date:
            # Check if end date is in the future
            from datetime import datetime
            if recurrence_data.end_date < datetime.now():
                return False
                
        return True

    def generate_next_occurrence_date(self, recurrence_pattern: RecurrencePattern, current_date: datetime) -> Optional[datetime]:
        """
        Calculate the next occurrence date based on the recurrence pattern
        """
        from datetime import timedelta
        
        if recurrence_pattern.frequency == "daily":
            return current_date + timedelta(days=recurrence_pattern.interval)
            
        elif recurrence_pattern.frequency == "weekly":
            # For weekly, we need to find the next occurrence based on days of week
            if recurrence_pattern.days_of_week:
                # This is a simplified implementation - a full implementation would need to handle
                # finding the next matching day of the week
                return current_date + timedelta(weeks=recurrence_pattern.interval)
                
        elif recurrence_pattern.frequency == "monthly":
            # For monthly, add the interval number of months
            # This is a simplified implementation - a full implementation would handle
            # month-end edge cases (e.g., Jan 31 + 1 month)
            import calendar
            year = current_date.year
            month = current_date.month + recurrence_pattern.interval
            
            # Handle year overflow
            while month > 12:
                year += 1
                month -= 12
                
            # Handle day overflow (e.g., Jan 31 -> Feb 31 doesn't exist)
            max_day = calendar.monthrange(year, month)[1]
            day = min(current_date.day, max_day)
            
            return current_date.replace(year=year, month=month, day=day)
            
        elif recurrence_pattern.frequency == "yearly":
            # For yearly, add the interval number of years
            return current_date.replace(year=current_date.year + recurrence_pattern.interval)
            
        return None

    def should_end_recurrence(self, recurrence_pattern: RecurrencePattern, occurrence_count: int, current_date: datetime) -> bool:
        """
        Determine if a recurrence should end based on its end conditions
        """
        if recurrence_pattern.end_condition == "never":
            return False
            
        elif recurrence_pattern.end_condition == "after":
            return occurrence_count >= recurrence_pattern.end_after_occurrences
            
        elif recurrence_pattern.end_condition == "on_date":
            from datetime import datetime
            return current_date > recurrence_pattern.end_date
            
        return False

    def validate_recurrence_pattern(self, recurrence_pattern: RecurrencePattern) -> bool:
        """
        Validate a recurrence pattern
        """
        # Validate frequency
        if recurrence_pattern.frequency not in ["daily", "weekly", "monthly", "yearly"]:
            return False
            
        # Validate interval
        if recurrence_pattern.interval < 1:
            return False
            
        # Validate days of week if frequency is weekly
        if recurrence_pattern.frequency == "weekly" and recurrence_pattern.days_of_week:
            valid_days = {"mon", "tue", "wed", "thu", "fri", "sat", "sun"}
            if not all(day in valid_days for day in recurrence_pattern.days_of_week):
                return False
                
        # Validate day of month if frequency is monthly
        if recurrence_pattern.frequency == "monthly" and recurrence_pattern.day_of_month:
            if not (1 <= recurrence_pattern.day_of_month <= 31):
                return False
                
        # Validate end condition
        if recurrence_pattern.end_condition not in ["never", "after", "on_date"]:
            return False
            
        # Validate end after occurrences if end condition is "after"
        if recurrence_pattern.end_condition == "after" and recurrence_pattern.end_after_occurrences:
            if recurrence_pattern.end_after_occurrences < 1:
                return False
                
        # Validate end date if end condition is "on_date"
        if recurrence_pattern.end_condition == "on_date" and recurrence_pattern.end_date:
            # Check if end date is in the future
            from datetime import datetime
            if recurrence_pattern.end_date < datetime.now():
                return False
                
        return True

File: frontend/src/services/recurrenceService.ts
import { RecurrencePattern } from '../models/recurrence';

export class RecurrenceService {
  /**
   * Generates future task instances based on a recurrence pattern
   * @param pattern The recurrence pattern to use
   * @param startDate The start date to begin generating instances from
   * @param endDate The end date to stop generating instances (optional)
   * @param maxInstances Maximum number of instances to generate (default: 100)
   * @returns Array of dates representing future task instances
   */
  static generateFutureInstances(
    pattern: RecurrencePattern,
    startDate: Date,
    endDate?: Date,
    maxInstances: number = 100
  ): Date[] {
    const instances: Date[] = [];
    let currentDate = new Date(startDate);
    
    // Adjust to the next occurrence based on the pattern
    currentDate = this.getNextOccurrence(pattern, currentDate);
    
    while (
      (!endDate || currentDate <= endDate) &&
      instances.length < maxInstances
    ) {
      instances.push(new Date(currentDate));
      
      // Calculate next occurrence based on pattern
      currentDate = this.getNextOccurrence(pattern, currentDate, true);
      
      // Check end conditions
      if (pattern.endCondition === 'after_occurrences' && instances.length >= (pattern.occurrenceCount || 0)) {
        break;
      }
    }
    
    return instances;
  }

  /**
   * Calculates the next occurrence based on the recurrence pattern
   * @param pattern The recurrence pattern
   * @param fromDate The date to calculate from
   * @param isIncrement Whether this is an increment from a previous occurrence
   * @returns The next occurrence date
   */
  private static getNextOccurrence(
    pattern: RecurrencePattern,
    fromDate: Date,
    isIncrement: boolean = false
  ): Date {
    const nextDate = new Date(fromDate);
    
    switch (pattern.patternType) {
      case 'daily':
        if (isIncrement) {
          nextDate.setDate(nextDate.getDate() + pattern.interval);
        }
        break;
        
      case 'weekly':
        if (isIncrement) {
          // Move to next week
          nextDate.setDate(nextDate.getDate() + 7 * pattern.interval);
          
          // If days of week are specified, find the next valid day
          if (pattern.daysOfWeek && pattern.daysOfWeek.length > 0) {
            return this.findNextValidDay(nextDate, pattern.daysOfWeek);
          }
        } else {
          // If days of week are specified, find the next valid day from fromDate
          if (pattern.daysOfWeek && pattern.daysOfWeek.length > 0) {
            return this.findNextValidDay(nextDate, pattern.daysOfWeek);
          }
        }
        break;
        
      case 'monthly':
        if (isIncrement) {
          // Add months based on interval
          nextDate.setMonth(nextDate.getMonth() + pattern.interval);
          
          // If specific days of month are specified, adjust to the next valid day
          if (pattern.daysOfMonth && pattern.daysOfMonth.length > 0) {
            return this.findNextValidDayOfMonth(nextDate, pattern.daysOfMonth);
          }
        } else {
          // If specific days of month are specified, adjust to the next valid day from fromDate
          if (pattern.daysOfMonth && pattern.daysOfMonth.length > 0) {
            return this.findNextValidDayOfMonth(nextDate, pattern.daysOfMonth);
          }
        }
        break;
        
      case 'yearly':
        if (isIncrement) {
          // Add years based on interval
          nextDate.setFullYear(nextDate.getFullYear() + pattern.interval);
        }
        break;
    }
    
    return nextDate;
  }

  /**
   * Finds the next valid day of the week from the specified date
   */
  private static findNextValidDay(fromDate: Date, validDays: Array<'mon' | 'tue' | 'wed' | 'thu' | 'fri' | 'sat' | 'sun'>): Date {
    const dayMap: { [key: string]: number } = {
      'sun': 0, 'mon': 1, 'tue': 2, 'wed': 3, 'thu': 4, 'fri': 5, 'sat': 6
    };
    
    const validDayNumbers = validDays.map(day => dayMap[day]);
    const fromDateDay = fromDate.getDay();
    
    // If the current day is a valid day and we're looking for the next occurrence, 
    // we need to find the next valid day
    if (validDayNumbers.includes(fromDateDay)) {
      // This is already a valid day, so return it
      return new Date(fromDate);
    }
    
    // Find the next valid day
    for (let i = 1; i <= 7; i++) {
      const nextDay = (fromDateDay + i) % 7;
      if (validDayNumbers.includes(nextDay)) {
        const nextDate = new Date(fromDate);
        nextDate.setDate(nextDate.getDate() + i);
        return nextDate;
      }
    }
    
    // Fallback (should not happen if validDays is not empty)
    return new Date(fromDate);
  }

  /**
   * Finds the next valid day of the month from the specified date
   */
  private static findNextValidDayOfMonth(fromDate: Date, validDays: number[]): Date {
    const currentDay = fromDate.getDate();
    
    // Find the next valid day in the current month
    for (const day of validDays) {
      if (day >= currentDay) {
        const nextDate = new Date(fromDate);
        nextDate.setDate(day);
        return nextDate;
      }
    }
    
    // If no valid day remains in the current month, 
    // find the first valid day in the next month
    const nextDate = new Date(fromDate);
    nextDate.setMonth(nextDate.getMonth() + 1);
    nextDate.setDate(Math.min(...validDays)); // Use the first valid day of the month
    
    return nextDate;
  }

  /**
   * Checks if a recurrence pattern would conflict with existing tasks
   * @param pattern The recurrence pattern to check
   * @param existingTaskDates Dates of existing tasks to check against
   * @returns True if there's a conflict, false otherwise
   */
  static checkForConflicts(
    pattern: RecurrencePattern,
    existingTaskDates: Date[],
    startDate: Date,
    endDate?: Date
  ): boolean {
    const futureInstances = this.generateFutureInstances(pattern, startDate, endDate);
    
    return futureInstances.some(instance => 
      existingTaskDates.some(existingDate => 
        this.isSameDay(instance, existingDate)
      )
    );
  }

  /**
   * Checks if two dates are on the same day
   */
  private static isSameDay(date1: Date, date2: Date): boolean {
    return (
      date1.getFullYear() === date2.getFullYear() &&
      date1.getMonth() === date2.getMonth() &&
      date1.getDate() === date2.getDate()
    );
  }
}